pkg/gstreamer/bin.go

```diff
@@ -0,0 +1,427 @@
+// Copyright 2023 LiveKit, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gstreamer
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/tinyzimmer/go-gst/gst"
+
+	"github.com/livekit/egress/pkg/errors"
+	"github.com/livekit/protocol/logger"
+)
+
+// Bins are designed to hold a single stream, with any number of sources and sinks
+type Bin struct {
```

Comment:
Looks like an useful abstraction! Some comments about general usage would be helpful for me personally. looks like this has potential of being in a livekit-gst package? (shared between our multiple apps)

Reply:
I did have that in mind when writing it, aiming do that after PC. Right now it has a couple egress-specific things like callbacks, and it needs more work to make sure (or at least explain how) it's used properly. Can definitely add more comments for now